tee_test.go

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zap
+
+import (
+	"bytes"
+	"errors"
+	"testing"
+
+	"github.com/uber-go/zap/spywrite"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestTeeWritesBoth(t *testing.T) {
+	first := &bytes.Buffer{}
+	second := &bytes.Buffer{}
+	ws := Tee(AddSync(first), AddSync(second))
+
+	msg := []byte("dumbledore")
+	n, err := ws.Write(msg)
+	require.NoError(t, err, "Expected successful buffer write")
+	assert.Equal(t, len(msg), n)
+
+	assert.Equal(t, msg, first.Bytes())
+	assert.Equal(t, msg, second.Bytes())
+}
+
+func TestTeeFailsWrite(t *testing.T) {
+	failer := spywrite.FailWriter{}
+	ws := Tee(AddSync(failer))
+
+	_, err := ws.Write([]byte("test"))
+	assert.Error(t, err, "Write error should propagate")
+}
+
+func TestTeeFailsShortWrite(t *testing.T) {
+	shorter := spywrite.ShortWriter{}
+	ws := Tee(AddSync(shorter))
+
+	n, err := ws.Write([]byte("test"))
+	assert.NoError(t, err, "Expected fake-success from short write")
+	assert.Equal(t, 3, n, "Expected byte count to return from underlying writer")
+}
+
+func TestWritestoAllSyns_EvenIfFirstErrors(t *testing.T) {
+	failer := spywrite.FailWriter{}
+	shortWriter := spywrite.ShortWriter{}
+	ws := Tee(AddSync(failer), AddSync(shortWriter))
+
+	n, err := ws.Write([]byte("fail"))
+	assert.Error(t, err, "Expected error from call to a writer that failed")
+	assert.Equal(t, 3, n, "Short writer should drop a byte")
```

Comment:
should we use a successful writer after the fail writer and verify contents instead? (this works too, just seems a little clearer to show that the second one can succeed)